File: js/packages/app/src/__tests__/features/object-references/external/polywrap/object-references-example/src/polywrap/external/global-functions/testInvokeExternalInstanceMethod.ts
import { concat, u32ToBuffer, IExternalWrapInstance, BaseTypeSerialization } from '@polywrap/reim-wrap-js';
import { WrapModule } from '../module/WrapModule';
import { WrapManifest } from '../../WrapManifest';
import { ExternalResource } from '../../dt/ExternalResource';


export function testInvokeExternalInstanceMethod(
  arg: string,
): Promise<string> {
  return testInvokeExternalInstanceMethodFromInstance(
    WrapModule.wrapInstance,
    arg,
  );
};

export const create = (instance: IExternalWrapInstance) => {
  return (
    arg: string,
  ): Promise<string> => {
    return testInvokeExternalInstanceMethodFromInstance(
      instance, 
      arg,
    );
  };
};

export const testInvokeExternalInstanceMethodFromInstance = async (
  instance: IExternalWrapInstance | null, 
  arg: string,
): Promise<string> => {
  if (instance == null) {
    throw new Error("connect() or import() must be called before using this module");
  }

  const args = new TestInvokeExternalInstanceMethodArgs(
    arg,
  );

  const buffer = concat([
    u32ToBuffer(WrapManifest.External.GlobalFunction.TestInvokeExternalInstanceMethod),
    TestInvokeExternalInstanceMethodArgsWrapped.serialize(args),
  ]);

  const result = await instance.invokeResource(ExternalResource.InvokeGlobalFunction, buffer);

  
  return BaseTypeSerialization.deserialize<string>(result);
}

export class TestInvokeExternalInstanceMethodArgsWrapped {
  constructor(
    public arg: string,
  ) {
  }

  static serialize(value: TestInvokeExternalInstanceMethodArgs): Uint8Array {
    return new TextEncoder().encode(
      JSON.stringify(
        TestInvokeExternalInstanceMethodArgsWrapped.mapToSerializable(value)
      )
    );
  }

  static mapToSerializable(value: TestInvokeExternalInstanceMethodArgs): TestInvokeExternalInstanceMethodArgsWrapped {
    return new TestInvokeExternalInstanceMethodArgsWrapped(
            
      value.arg,
      
    );
  }
}

export class TestInvokeExternalInstanceMethodArgs {
  constructor(
    public arg: string,
  ) {
  }
}

File: rust/app/src/polywrap/external_references/src/polywrap/external/global-functions/testInvokeExternalInstanceMethod.rs
use reim_wrap::{ ExternalWrapModule };
use module::{ WrapManifest };
use dt::{ ExternalResource };
use external_module::{ WrapModule };
use wrapped::{ testInvokeExternalInstanceMethodWrapped };


pub async fn testInvokeExternalInstanceMethod(
  arg: string,
) -> string {
  testInvokeExternalInstanceMethodFromInstance(
    WrapModule.wrapInstance,
    arg,
  )
};

pub fn create(instance: dyn ExternalWrapModule) {
  |
    arg: string,
  | {
    testInvokeExternalInstanceMethodFromInstance(
      instance, 
      arg,
    )
  }
}

pub async fn testInvokeExternalInstanceMethodFromInstance (
  instance: Option<ExternalWrapModule>, 
  arg: string,
) -> string {
  if instance.is_none() {
    panic!("connect() or import() must be called before using this module");
  }

  let args = TestInvokeExternalInstanceMethodArgs.new(
    arg,
  );

  let buffer = [
    WrapManifest.External.GlobalFunction.TestInvokeExternalInstanceMethod.to_be_bytes(),
    TestInvokeExternalInstanceMethodArgsWrapped.serialize(args),
  ].concat();

  let result = instance.invokeResource(ExternalResource.InvokeGlobalFunction, buffer).await;

  
  BaseTypeSerialization.deserialize<string>(result)
}

struct TestInvokeExternalInstanceMethodArgsWrapped {
    pub arg: string,
}
impl TestInvokeExternalInstanceMethodArgsWrapped {
    pub fn new(
      arg: string,
    ) -> TestInvokeExternalInstanceMethodArgsWrapped {
      TestInvokeExternalInstanceMethodArgsWrapped {
        arg,
      }
    }

    pub fn serialize(value: TestInvokeExternalInstanceMethodArgs) -> Vec<u8> {
        json!(
            TestInvokeExternalInstanceMethodArgsWrapped.mapToSerializable(value)
        ).as_bytes()
    }

    pub fn mapToSerializable(value: TestInvokeExternalInstanceMethodArgs) -> TestInvokeExternalInstanceMethodArgsWrapped {
        TestInvokeExternalInstanceMethodArgsWrapped.new(
                        
            value.arg,
            
        )
    }
}

struct TestInvokeExternalInstanceMethodArgs {
    pub arg: string,
}

impl TestInvokeExternalInstanceMethodArgs {
  pub fn new(
    arg: string,
  ) -> TestInvokeExternalInstanceMethodArgs {
    TestInvokeExternalInstanceMethodArgs.new (
      arg,
    )
  }
}
